File: src/java/eu/sig/training/ch02/BalancesServlet.java
package eu.sig.training.ch02;

import java.io.IOException;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Properties;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

public class BalancesServlet extends HttpServlet {
    private static final long serialVersionUID = 1L;
    private Properties conf;

    @Override
    // tag::doGet[]
    public void doGet(HttpServletRequest req, HttpServletResponse resp)
        throws ServletException, IOException {
        resp.setContentType("application/json");
        try {
            Connection conn = DriverManager.
                getConnection(this.conf.getProperty("handler.jdbcurl"));
            ResultSet results =
                conn.createStatement()
                    .executeQuery(
                        "SELECT account, balance FROM ACCTS WHERE id="
                            + req.getParameter(conf.
                                getProperty("request.parametername")));
            float totalBalance = 0;
            resp.getWriter().print("{\"balances\":[");
            while (results.next()) {
                // Assuming result is 9-digit bank account number,
                // validate with 11-test:
                int sum = 0;
                for (int i = 0; i < results.getString("account")
                    .length(); i++) {
                    sum = sum + (9 - i)
                            * Character.getNumericValue(results.getString(
                            "account").charAt(i));
                }
                if (sum % 11 == 0) {
                    totalBalance += results.getFloat("balance");
                    resp.getWriter().print(
                        "{\"" + results.getString("account") + "\":"
                            + results.getFloat("balance") + "}");
                }
                if (results.isLast()) {
                    resp.getWriter().println("],");
                } else {
                    resp.getWriter().print(",");
                }
            }
            resp.getWriter().println("\"total\":" + totalBalance + "}");
        } catch (SQLException e) {
            System.out.println("SQL exception: " + e.getMessage());
        }
    }
    // end::doGet[]
}

File: src/csharp/eu/sig/training/ch02/BalancesServlet.cs
using System;
using System.Web;
using System.Data;
using System.Data.SqlClient;
using System.Configuration;

namespace eu.sig.training.ch02
{
    public class BalancesServlet
    {
        // Visual Studio Code Analysis is right in pointing out that the following method has
        // security flaws. However, that's beside the point for this example.
        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Security", "CA2100:Review SQL queries for security vulnerabilities")]
        // tag::doGet[]
        public void DoGet(HttpRequest req, HttpResponse resp)
        {
            resp.ContentType = "application/json";
            string command = "SELECT account, balance " +
                "FROM ACCTS WHERE id=" + req.Params[
                    ConfigurationManager.AppSettings["request.parametername"]];
            SqlDataAdapter dataAdapter = new SqlDataAdapter(command,
                ConfigurationManager.AppSettings["handler.serverstring"]);
            DataSet dataSet = new DataSet();
            dataAdapter.Fill(dataSet, "ACCTS");
            DataTable dataTable = dataSet.Tables[0];
            try
            {
                float totalBalance = 0;
                int rowNum = 0;
                resp.Write("{\"balances\":[");
                while (dataTable.Rows.GetEnumerator().MoveNext())
                {
                    rowNum++;
                    DataRow results = (DataRow)dataTable.Rows.GetEnumerator().Current;
                    // Assuming result is 9-digit bank account number,
                    // validate with 11-test:
                    int sum = 0;
                    for (int i = 0; i < ((string)results["account"]).Length; i++)
                    {
                        sum = sum + (9 - i) *
                            (int)Char.GetNumericValue(((string)results["account"])[i]);
                    }
                    if (sum % 11 == 0)
                    {
                        totalBalance += (float)results["balance"];
                        resp.Write($"{{\"{results["account"]}\":{results["balance"]}}}");
                    }
                    if (rowNum == dataTable.Rows.Count)
                    {
                        resp.Write("],\n");
                    }
                    else
                    {
                        resp.Write(",");
                    }
                }
                resp.Write($"\"total\":{totalBalance}}}\n");
            }
            catch (SqlException e)
            {
                Console.WriteLine($"SQL exception: {e.Message}");
            }
        }
        // end::doGet[]
    }
}
